gameserver/New/Game.ts

```diff
@@ -0,0 +1,189 @@
+    private currentPlayerIndex = 0;
+    
+    private scoreTeam1 = 0; // 0 with 2
+    private scoreTeam2 = 0; // 1 with 3
+    private maxScore = 0;
+
+    private playedCards = new Array<CARDS>();
+    private firstTrump = true;
+    private currentTrumpColor = CARD_COLOR.SPADES;
+
+    public constructor(players: Player[], maxScore = 1000)
+    {
+        this.players = players; 
+        this.currentPlayerIndex = 0; 
+        this.maxScore = maxScore;
+    }
+
+    async wait(timems: number): Promise<void>
+    {
+        return new Promise((s, r) => {
+            setTimeout(() => {
+                s();
+            }, timems);
+        });
+    }
+    
+    private roomBroadcast(event: string, ...args: any[])
+    {
+        this.players.forEach(p => p.getSocket().emit(event, ...args));
+    }
+
+    async playGame()
+    {
+        for(let i = 0; i < this.players.length; ++i)
+        {
+            this.players[i].emitID(i);
+        }
+
+        for(let i = 0; i < this.players.length; ++i)
+        {
+            this.roomBroadcast("player", i, this.players[i].getName());
+        }
+
+        this.roomBroadcast("gameStart");
+
+        let trumpMakerId = 0;
+
+        while(this.scoreTeam1 < this.maxScore && this.scoreTeam2 < this.maxScore)
+        {
+            this.currentPlayerIndex = trumpMakerId;
+            await this.playRound();
+            this.currentPlayerIndex = trumpMakerId;
+            this.nextPlayer();
+            trumpMakerId = this.currentPlayerIndex;
+        }
+    }
+    
+    private getCurrentPlayer()
+    {
+        return this.players[this.currentPlayerIndex];
+    }
+
+    addTeamScoreOfPlayer(playerIndex: number, score: number)
+    {
+        if(playerIndex % 2 == 0)
+        {
+            this.scoreTeam1 += score;
+        }
+        else
+        {
+            this.scoreTeam2 += score;
+        }
+    }
+
+    async playRound()
+    {
+        // Distribute new cards
+        let deck = new Deck();
+        deck.shuffleDeck();
+        let cardNumber = deck.getDeck().length / this.players.length;
+        this.players.forEach(player => {
+            player.setCards(deck.getDeck().splice(0, cardNumber));
+        });                
+
+        // First round, player that has seven of diamonds chooses trump
+        if(this.firstTrump)
+        {
+            this.firstTrump = false;
+            this.players.forEach((player, index) => {
+                if(player.getCards().includes(CARDS.SEVEN_DIAMOND))
+                {
+                    this.currentPlayerIndex = index;
+                }
+            });
+        }
+
+        // Choose trump
+        try
+        {
+            this.currentTrumpColor = await this.getCurrentPlayer().chooseTrump();
+        }
+        catch
+        {
+            this.roomBroadcast("playerPassed", this.currentPlayerIndex);
+            this.currentTrumpColor = await this.players[(this.currentPlayerIndex + 2) % 4].chooseTrump(true);
```

Comment:
the % 4 here looks like a magic number. Isn't it better to use the player array size ?